### ISSUE TITLE
Determine best approach for working with Java apps

### ISSUE THREAD

**Issue by notmandatory** [bitcoindevkit/bdk-kotlin#47]
Reported by SeverinAlexB on discord. See also:

https://github.com/SeverinAlexB/bdk-java-ulong-bug

**Comment by thunderbiscuit** [comment]
@SeverinAlexB

**Comment by artfuldev** [comment]
I think ULong is the stricter type-safe API for Kotlin consumers but Java doesn't have the concept of unsigned numbers. I'm not sure if we're just targeting Kotlin consumers though. The way to solve this in a Java project is to just use Kotlin's libraries (which are JVM compatible) so we can either make that part of our documentation, or decide to weaken our type-safety in the APIs.

**Comment by thunderbiscuit** [comment]
Yeah I'm not sure where the bug is coming from, but for some reason this Kotlin/Java compatibility breaks here for `ULong`. Also of interest (not sure how related) is the fact that on the Java side, the `ElectrumConfig` is private (even though it's a public data class).

Keeping these here for archives (issue was discussed on the #ffi-bindings channel a bit more):

If you use incorrect syntax on the Java side, the IDE will suggest something, and the tooltip box shows a public data class:
<img width="484" alt="public" src="https://user-images.githubusercontent.com/39974688/161852169-0f87ecb1-3267-4485-bc72-54ee372a6708.png">

If you accept the changes proposed (adding the `new` keyword), now you get an error telling you the constructor for `ElectrumConfig` is private:
<img width="570" alt="private" src="https://user-images.githubusercontent.com/39974688/161852253-6cdff372-0707-4d26-a184-fa4401de19d1.png">

**Comment by thunderbiscuit** [comment]
Notice in the second picture the type for `timeout` is `kotlin.UByte`, whereas the type for `retry` and `stopGap` are `byte` and `long` respectively...

**Comment by SeverinAlexB** [comment]
Hey guys, thanks for taking this on

I saw another issue I can't explain. For some reason, I can't call methods that are available in Kotlin. For example, `sync` or `getBalance` are just not available. As always, it works with a custom Kotlin wrapper.
![image](https://user-images.githubusercontent.com/8782386/161924106-894c5a8f-d37a-4929-9ffc-8b10ebae501a.png)

All methods that are available in Java:
<img width="595" alt="CleanShot 2022-04-06 at 11 02 45@2x" src="https://user-images.githubusercontent.com/8782386/161926184-76e92a4e-57d8-422c-9aa3-01e5b031f57c.png">

**Comment by artfuldev** [comment]
I'll take a deeper look at this, thanks!

**Comment by thunderbiscuit** [comment]
Will attempt to keep track of the known issues and some hacks/fixes to get around them for the time being [here](https://hackmd.io/@thunderbiscuit/HJx0a5g1Sc/%2Fe-oqaAh8RK6ax8Ei5w8uWw).
